source/grid/grid_tools.cc

```diff
@@ -1458,6 +1458,145 @@ namespace GridTools
+    for (unsigned int vertex=0; vertex<n_vertices; ++vertex)
+      if (mesh.vertex_used(vertex))
+        {
+          const unsigned int n_neighbor_cells = vertex_to_cells[vertex].size();
+          vertex_to_cell_centers[vertex].resize(n_neighbor_cells);
+
+          typename std::set<typename Triangulation<dim,spacedim>::active_cell_iterator>::iterator it = vertex_to_cells[vertex].begin();
+          for (unsigned int cell=0; cell<n_neighbor_cells; ++cell,++it)
+            {
+              vertex_to_cell_centers[vertex][cell] = (*it)->center() - vertices[vertex];
+              vertex_to_cell_centers[vertex][cell] /= vertex_to_cell_centers[vertex][cell].norm();
+            }
+        }
+    return vertex_to_cell_centers;
+  }
+
+
+  namespace
+  {
+    template <int spacedim>
+    bool
+    compair_point_association(const unsigned int a,
+                              const unsigned int b,
+                              const Tensor<1,spacedim> &point_direction,
+                              const std::vector<Tensor<1,spacedim> > &center_directions)
+    {
+      const double scalar_product_a = center_directions[a] * point_direction;
+      const double scalar_product_b = center_directions[b] * point_direction;
+
+      // The function is supposed to return if a is before b. We are looking
+      // for the alignment of particle direction and center direction,
+      // therefore return if the scalar product of a is larger.
+      return (scalar_product_a > scalar_product_b);
+    }
+  }
+
+  template <int dim, template <int, int> class MeshType, int spacedim>
+#ifndef _MSC_VER
+  std::pair<typename MeshType<dim, spacedim>::active_cell_iterator, Point<dim> >
+#else
+  std::pair<typename dealii::internal::ActiveCellIterator<dim, spacedim, MeshType<dim, spacedim> >::type, Point<dim> >
+#endif
+  find_active_cell_around_point (const Mapping<dim,spacedim>                                                    &mapping,
+                                 const MeshType<dim,spacedim>                                                   &mesh,
+                                 const Point<spacedim>                                                          &p,
+                                 const typename MeshType<dim, spacedim>::active_cell_iterator                   &current_cell,
+                                 const std::vector<std::set<typename MeshType<dim,spacedim>::active_cell_iterator > > &vertex_to_cells,
+                                 const std::vector<std::vector<Tensor<1,spacedim> > >                            &vertex_to_cell_centers,
+                                 const std::vector<bool>                                                         &marked_vertices)
+  {
+    std::pair<typename MeshType<dim, spacedim>::active_cell_iterator, Point<dim> > cell_and_position;
+
+    // Create a corresponding map of vectors from vertex to cell center
+    std::vector<unsigned int> neighbor_permutation;
```

Comment:
This line can be moved and merged with line 1534 as `std::vector<unsigned int> neighbor_permutation(n_neighbor_cells)`. In ASPECT I repeat the steps below for many points, that is why I split the declaration from the use, but this is not longer necessary here.